receiver/dockerstatsreceiver/logs_receiver.go

```diff
@@ -0,0 +1,201 @@
+		// except for context cancellation.
+		eventChan, errChan := d.client.Events(ctx, events.ListOptions{
+			Since:   d.config.Logs.Since,
+			Until:   d.config.Logs.Until,
+			Filters: filterArgs,
+		})
+
+		err := d.processEvents(ctx, eventChan, errChan)
+		if errors.Is(err, context.Canceled) {
+			return
+		}
+		// io.EOF is expected, no need to log but still needs retrying
+		if err != nil && !errors.Is(err, io.EOF) && !errors.Is(err, io.ErrUnexpectedEOF) {
+			d.logger.Error("Async error while processing docker events, reconnecting to docker daemon", zap.Error(err))
+		}
+		nextBackoff := d.backoff.NextBackOff()
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(nextBackoff):
+			continue
+		}
+	}
+}
+
+func (d *dockerEventPoller) processEvents(ctx context.Context, eventChan <-chan events.Message, errChan <-chan error) error {
+	processedOnce := false
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case event := <-eventChan:
+			// for the given method invocation, processing event indicates a successful daemon connection.
+			// backoff should be reset, no need to do this afterwards since the connection is already established.
+			if !processedOnce {
+				d.backoff.Reset()
+				processedOnce = true
+			}
+			if err := d.eventHandler(ctx, &event); err != nil {
+				d.logger.Error("Failed to process docker event", zap.Error(err))
+			}
+		case err := <-errChan:
+			return err
+		}
+	}
+}
+
+// TODO: add batching based on time/volume. for now, one event -> one logs
+func (r *logsReceiver) consumeDockerEvent(ctx context.Context, event *events.Message) error {
+	if event.Type == "" {
+		return nil
+	}
+	logs := plog.NewLogs()
+	logRecord := logs.ResourceLogs().AppendEmpty().ScopeLogs().AppendEmpty().LogRecords().AppendEmpty()
+	logRecord.SetTimestamp(pcommon.NewTimestampFromTime(time.Unix(event.Time, event.TimeNano)))
+	attrs := logRecord.Attributes()
+	action := string(event.Action)
+	// for cases with health_status: running etc., event.name should be
+	// properly namespaced as docker.container.health_status.running
+	if strings.Contains(action, ": ") {
+		action = strings.Join(strings.Split(action, ": "), ".")
+	}
+	if action != "" {
+		// i.e. docker.container.start
+		attrs.PutStr("event.name", fmt.Sprintf("%s.%s.%s", dockerEventPrefix, event.Type, action))
+	} else {
+		attrs.PutStr("event.name", fmt.Sprintf("%s.%s", dockerEventPrefix, event.Type))
+	}
+	if event.Scope != "" {
+		attrs.PutStr("event.scope", event.Scope)
+	}
+	if event.Actor.ID != "" {
+		attrs.PutStr("event.id", event.Actor.ID)
+	}
+	// body exactly replicates actor attributes
+	if len(event.Actor.Attributes) > 0 {
+		actorAttrs := logRecord.Body().SetEmptyMap()
+		for k, v := range event.Actor.Attributes {
+			if k != "" {
+				actorAttrs.PutStr(k, v)
+			}
+		}
+	}
+	return r.consumer.ConsumeLogs(ctx, logs)
+}
+
+func (r *logsReceiver) Shutdown(_ context.Context) error {
+	if r.cancel != nil {
```

Comment:
Consider a `sync.WaitGroup` to ensure the goroutine terminates synchronously with `Shutdown()`

Reply:
added